### PR TITLE
Remove the need to wrap views in 'withSSRData'

### PR DESCRIPTION
**Pull request by POORYAJ**
First of all, sorry for opening a pull request since you guys mentioned open an issue first, but because I wanted there to be a working example of what's being discussed rather than talking about it abstractly I opened a PR. But there is no obligation on you guys and it's ok you if you want to dismiss the work that has been done.

## Main Changes

I think the need for wrapping the views in `withSSRData` is an unnecessary and extra step, and may lead to bugs since you may forget to add it. It also can be unfamiliar to other people since other frameworks don't use HOCs to pass the server side props.
Since the logic of it is necessary for views, I used the same component but moved the duty of wrapping the view in it from component level to the route level .

This is possible by implementing the same exact api and functionality as `renderRoutes` that is provided by `react-router-config` but with the extra logic for wrapping components in `withSSRData` if they've used `serverSideInitial`. This change is also compatible with the old way of writing views that are explicitly wrapped in `withSSRData`.

Because wrapping the view is done implicitly in the route level now, the old logic for `hasPreloadedData` that was returned from `findFinalComponent` doesn't work because it relied on if the component is wrapped in `withSSRData` explicitly and now that the need for `withSSRData` is gone the implicit components aren't wrapped at the moment of checking them, so I've changed the logic for `hasPreloadedData` to check for the return value of `serverSideInitial`, since if it return it means the view needs the data, and if it doesn't return anything we can set `hasPreloadedData` as false. And to remain backward compatible it also checks for explicit usage of `withSSRData` too.

But I think it's better to remove this check altogether and make both the code and the mental model easier to understand: if you use `serverSideInitial` you automatically get a `initialSSRData` prop that if you return something in your function it's available there but if you don't, it's an empty object and you can ignore it.

## Bug Fix

In the process I also fixed a bug related to using render props for routes that required you to pass both the `component` and `render` properties in the route config to make it work. This bug was caused by `getInitialPropsList` since it was expecting that components only come from the `component` property and not the `render` property.

so basically with the fix you can make routes like this:
```diff
{
- component: Home,
  path: '/',
  render: (props) => <Home {...props} />,
}
```

___

Another possible solution for removing `withSSRData` is to make a babel plugin to wrap the views in `withSSRData` implicitly. An advantage is that it makes the code behave as it was without any further changes, but I chose not to do it that way since it makes the components rely on babel and make further transition to the upcoming toolings (like vite) hard and may make further contributors a little confused about how this stuff works internally.

I would like to hear your ideas too and I'm open to discuss any further changes.

### CODE REVIEW

File: src/server/utils.js [hunk 1]
```diff
@@ -1,5 +1,10 @@
+import React from 'react';
+import { Switch, Route } from 'react-router-dom';
+
 import { getEnv } from 'src/utils/env';
 
+import withSSRData from '../shared-components/withSSRData';
```
Comment (iMohammadReza) [inline src/server/utils.js]: So `withSSRData` can be move from `shared-components` to `src/services/initial-ssr-data/`

**Comment by iMohammadReza** [comment]
Hi dear Poorya,
Thanks for your PR and your nice contribution.
I update doc, lint your changes and made some small refactor.
Also, Sorry my late.
Best regards.
